File: src/lib/renderOutput.ts
// What the final render is ENCODED as — container, video codec, quality, audio.
//
// Every one of these was hardcoded in worker/handlers/render.py until now:
// libx264 / yuv420p / crf 18 / veryfast, aac 192k 48kHz stereo, always .mp4.
// Fine as a default and useless as the only option — a cut going to an editor
// wants ProRes, a cut going on the web wants VP9, and a 90-second master at
// crf 18 is not the same file as one at crf 28.
//
// THE TABLE IS THE CONTRACT AND IT IS TWINNED with worker/render_output.py.
// This side decides what is OFFERED and whether a combination is legal; the
// worker builds the ffmpeg arguments. Only the worker encodes, so there is no
// second implementation of the encode to drift — but a format this file offers
// and the worker cannot build is a render that dies after the GPU time is
// spent, so `renderOutput.test.ts` parses the Python and fails if the two
// tables disagree.
//
// EVERY CODEC HERE WAS CHECKED AGAINST THE POD'S OWN ffmpeg (4.4.2) on
// 2026-08-23 — `ffmpeg -encoders` plus an NVENC smoke render. Nothing is
// offered on the strength of it usually being available. AV1 is absent for
// exactly that reason: this build has neither libsvtav1 nor av1_nvenc.
//
// Dependency-free, same rule as postChain.ts — db/types.ts imports from here.

export type VideoFormatId = "h264" | "h265" | "prores" | "vp9";
export type AudioCodecId = "aac" | "opus" | "pcm_s24le" | "none";

export interface VideoFormatDef {
  id: VideoFormatId;
  label: string;
  /** File extension AND the container ffmpeg infers from it. */
  ext: "mp4" | "mov" | "webm";
  contentType: string;
  /** Software encoder. */
  encoder: string;
  /** NVENC equivalent, where one exists on this pod. Absent = no hardware
   *  path, and the Hardware switch is hidden rather than ignored. */
  hwEncoder?: string;
  /** How quality is expressed. `crf` is a number that gets LOWER as quality
   *  rises; `profile` is ProRes's enum, which does not work that way at all —
   *  conflating them is how a "quality 3" slider ends up meaning 422 LT. */
  quality: "crf" | "profile";
  qualityMin: number;
  qualityMax: number;
  qualityDefault: number;
  /** Audio codecs this container will actually mux. */
  audio: AudioCodecId[];
  hint: string;
}

/** Ordered as the picker shows them: delivery first, then handoff. */
export const VIDEO_FORMATS: VideoFormatDef[] = [
  {
    id: "h264", label: "H.264", ext: "mp4", contentType: "video/mp4",
    encoder: "libx264", hwEncoder: "h264_nvenc",
    quality: "crf", qualityMin: 14, qualityMax: 32, qualityDefault: 18,
    audio: ["aac", "none"],
    hint: "Plays everywhere. The default, and the right answer unless you have "
        + "a reason — every browser, phone and NLE reads it.",
  },
  {
    id: "h265", label: "H.265 / HEVC", ext: "mp4", contentType: "video/mp4",
    encoder: "libx265", hwEncoder: "hevc_nvenc",
    quality: "crf", qualityMin: 18, qualityMax: 36, qualityDefault: 22,
    audio: ["aac", "none"],
    hint: "Roughly half the size of H.264 at the same look. Playback support "
        + "is good but not universal — Safari and modern NLEs yes, older "
        + "Windows players no. Note its CRF scale is not H.264's: 22 here is "
        + "about 18 there.",
  },
  {
    id: "prores", label: "ProRes 422", ext: "mov", contentType: "video/quicktime",
    encoder: "prores_ks",
    quality: "profile", qualityMin: 0, qualityMax: 5, qualityDefault: 3,
    audio: ["pcm_s24le", "aac", "none"],
    hint: "Editorial handoff — intra-frame, effectively lossless at HQ, and "
        + "what a colourist or an NLE wants. Very large: budget roughly 1GB "
        + "per minute at 1080p HQ.",
  },
  {
    id: "vp9", label: "VP9", ext: "webm", contentType: "video/webm",
    encoder: "libvpx-vp9",
    quality: "crf", qualityMin: 24, qualityMax: 45, qualityDefault: 31,
    audio: ["opus", "none"],
    hint: "Web delivery where you control the player. Smaller than H.264, and "
        + "slow to encode — there is no hardware encoder for it.",
  },
];

export const VIDEO_FORMAT = Object.fromEntries(
  VIDEO_FORMATS.map((f) => [f.id, f]),
) as Record<VideoFormatId, VideoFormatDef>;

/** ProRes `profile` is an enum, not a scale — named so the slider can say what
 *  it is picking rather than showing a bare 0-5. */
export const PRORES_PROFILES: Record<number, string> = {
  0: "Proxy", 1: "LT", 2: "422", 3: "422 HQ", 4: "4444", 5: "4444 XQ",
};

export interface AudioCodecDef {
  id: AudioCodecId;
  label: string;
  /** Bitrate is meaningless for PCM (it is uncompressed) and for none. */
  bitrates?: number[];
  bitrateDefault?: number;
  hint: string;
}

export const AUDIO_CODECS: AudioCodecDef[] = [
  { id: "aac", label: "AAC", bitrates: [128, 192, 256, 320], bitrateDefault: 192,
    hint: "The usual choice for a delivery file." },
  { id: "opus", label: "Opus", bitrates: [96, 128, 192, 256], bitrateDefault: 128,
    hint: "Better than AAC at the same bitrate, and what WebM carries." },
  { id: "pcm_s24le", label: "PCM 24-bit",
    hint: "Uncompressed, for editorial handoff. Large, and only in a .mov." },
  { id: "none", label: "No audio",
    hint: "Strips the track entirely — a picture-only master." },
];

export const AUDIO_CODEC = Object.fromEntries(
  AUDIO_CODECS.map((a) => [a.id, a]),
) as Record<AudioCodecId, AudioCodecDef>;

export const SAMPLE_RATES = [44100, 48000];

export interface RenderOutput {
  format: VideoFormatId;
  /** CRF for the crf formats, ProRes profile index for ProRes. One field
   *  because it is one control; `VideoFormatDef.quality` says how to read it. */
  quality: number;
  /** Use the NVENC encoder where the format has one. Ignored (and the control
   *  hidden) where it does not. */
  hardware: boolean;
  audio: AudioCodecId;
  audioBitrate: number;
  sampleRate: number;
  /** Delivery resolution as the SHORT EDGE in pixels, or null for the
   *  timeline's own frame.
   *
   *  The short edge and not the height, because that is the one reading that
   *  works both ways up: 1080p is ?x1080 on a landscape cut and 1080x? on a
   *  vertical one, and both are what "1080p" means to the person asking. The
   *  timeline's aspect is preserved either way — this resizes the frame, it
   *  never re-frames it, so a 1280x704 cut at 1080p delivers 1964x1080 rather
   *  than a letterboxed 1920x1080. The modal shows the computed size, so an
   *  aspect that is not 16:9 is visible rather than surprising. */
  resolution: number | null;
  /** Output frame rate, or null for the timeline's own. */
  fps: number | null;
}

export const RENDER_OUTPUT_DEFAULT: RenderOutput = {
  format: "h264",
  quality: 18,
  // Software by default even though NVENC is measurably faster: x264 at crf 18
  // is the file this studio has always produced, and a master that silently
  // changed encoder would change what every past render is compared against.
  hardware: false,
  audio: "aac",
  audioBitrate: 192,
  sampleRate: 48000,
  resolution: null,
  fps: null,
};

/** Short-edge targets, with the label people actually say. 540p earns its
 *  place as the review/proxy size; 4K is 2160 because that is what everyone
 *  outside a DCI house means by it (DCI 4K is 4096 wide and is not on offer). */
export const RESOLUTIONS: { value: number; label: string; title: string }[] = [
  { value: 540, label: "540p", title: "540 on the short edge — a review or proxy copy" },
  { value: 720, label: "720p", title: "720 on the short edge" },
  { value: 1080, label: "1080p", title: "1080 on the short edge" },
  { value: 1440, label: "1440p", title: "1440 on the short edge (QHD)" },
  { value: 2160, label: "4K", title: "2160 on the short edge (UHD)" },
];

const clamp = (n: number, lo: number, hi: number) => Math.min(hi, Math.max(lo, n));

/** Read whatever is on the row into a usable spec.
 *
 *  Forgiving where the value is merely out of range (clamp) and CORRECTIVE
 *  where the combination is illegal (an audio codec the container cannot mux
 *  falls back to that container's first). The alternative — passing it through
 *  — is a render that dies in ffmpeg after every clip has already been
 *  encoded, which on a long cut is half an hour of GPU time to learn that
 *  webm will not carry AAC. */
export function normalizeOutput(v: unknown): RenderOutput {
  const o = (v && typeof v === "object" && !Array.isArray(v) ? v : {}) as Partial<RenderOutput>;
  const fmt = VIDEO_FORMAT[o.format as VideoFormatId] ? (o.format as VideoFormatId)
                                                      : RENDER_OUTPUT_DEFAULT.format;
  const def = VIDEO_FORMAT[fmt];
  const quality = clamp(
    Number.isFinite(o.quality) ? Math.round(Number(o.quality)) : def.qualityDefault,
    def.qualityMin, def.qualityMax);
  const audio: AudioCodecId = def.audio.includes(o.audio as AudioCodecId)
    ? (o.audio as AudioCodecId) : def.audio[0];
  const acodec = AUDIO_CODEC[audio];
  const rates = acodec.bitrates;
  const audioBitrate = rates
    ? (rates.includes(Number(o.audioBitrate)) ? Number(o.audioBitrate) : acodec.bitrateDefault!)
    : 0;
  const sampleRate = SAMPLE_RATES.includes(Number(o.sampleRate))
    ? Number(o.sampleRate) : RENDER_OUTPUT_DEFAULT.sampleRate;
  const resolution = RESOLUTIONS.some((r) => r.value === Number(o.resolution))
    ? Number(o.resolution) : null;
  const fps = Number.isFinite(o.fps as number) && Number(o.fps) > 0
    ? clamp(Math.round(Number(o.fps)), 1, 120) : null;
  return {
    format: fmt, quality,
    hardware: def.hwEncoder ? o.hardware === true : false,
    audio, audioBitrate, sampleRate, resolution, fps,
  };
}

/** The encoder that will actually run — what the summary line should name, so
 *  "Hardware" is never a switch whose effect you have to infer. */
export function encoderOf(o: RenderOutput): string {
  const def = VIDEO_FORMAT[o.format];
  return o.hardware && def.hwEncoder ? def.hwEncoder : def.encoder;
}

/** "H.264 · CRF 18 · AAC 192k · .mp4" — the closed-state summary. */
export function describeOutput(o: RenderOutput): string {
  const def = VIDEO_FORMAT[o.format];
  const q = def.quality === "profile"
    ? `ProRes ${PRORES_PROFILES[o.quality] ?? o.quality}`
    : `CRF ${o.quality}`;
  const a = o.audio === "none" ? "no audio"
    : AUDIO_CODEC[o.audio].bitrates ? `${AUDIO_CODEC[o.audio].label} ${o.audioBitrate}k`
    : AUDIO_CODEC[o.audio].label;
  const parts = [def.label, q, a];
  if (o.hardware && def.hwEncoder) parts.push("hardware");
  if (o.resolution) parts.push(RESOLUTIONS.find((r) => r.value === o.resolution)!.label);
  if (o.fps) parts.push(`${o.fps}fps`);
  return `${parts.join(" · ")} · .${def.ext}`;
}

/** The delivered frame size, given the timeline's.
 *
 *  Even dimensions, because yuv420p cannot represent an odd one and ffmpeg
 *  FAILS the encode rather than rounding for you — on the last step of a long
 *  render. */
export function outputSize(o: RenderOutput, width: number, height: number): { w: number; h: number } {
  const even = (n: number) => Math.max(2, Math.round(n / 2) * 2);
  if (!o.resolution) return { w: even(width), h: even(height) };
  const f = o.resolution / Math.min(width, height);
  return { w: even(width * f), h: even(height * f) };
}

/** Is this target bigger than the cut actually is? The one question a named
 *  resolution invites you not to ask — "4K" reads like a quality setting and
 *  is really a resampling instruction. */
export function isUpscale(o: RenderOutput, width: number, height: number): boolean {
  return !!o.resolution && o.resolution > Math.min(width, height);
}

/** Warnings worth showing BEFORE a render that costs GPU minutes. Measured or
 *  structural only — no taste. */
export function outputWarnings(o: RenderOutput, width: number, height: number): string[] {
  const def = VIDEO_FORMAT[o.format];
  const out: string[] = [];
  const { w, h } = outputSize(o, width, height);
  if (isUpscale(o, width, height)) {
    out.push(`This cut is ${width}x${height}. Delivering ${w}x${h} resamples the `
           + `finished frame and adds no detail — the Upscale post pass is the `
           + `one that restores it.`);
  }
  if (o.format === "prores") {
    const mbPerMin = Math.round((w * h) / (1920 * 1080) * 880);
    out.push(`ProRes is large: roughly ${mbPerMin}MB per minute at ${w}x${h}.`);
  }
  if (o.format === "vp9") {
    out.push("VP9 has no hardware path here, so this encode is several times "
           + "slower than H.264.");
  }
  if (o.hardware && def.hwEncoder) {
    out.push("NVENC is much faster and gives up some quality per bit. For a "
           + "master, prefer software; for a review copy, this is the one to use.");
  }
  return out;
}

File: worker/render_output.py
"""What the final render is ENCODED as — the ffmpeg side of the output spec.

Python twin of src/lib/renderOutput.ts. Same format table, same normalisation
rules, same defaults; this side additionally BUILDS the arguments, because the
worker is the only thing that encodes. `src/lib/renderOutput.test.ts` parses
this file and fails if the two tables disagree — a format the browser offers
and this cannot build is a render that dies after every clip has been encoded.

Pure: no imports, no I/O, so it is testable off the pod.

Every codec here was checked against the pod's own ffmpeg (4.4.2) rather than
assumed available. AV1 is deliberately absent — that build has neither
libsvtav1 nor av1_nvenc.
"""

#: id -> (ext, content_type, encoder, hw_encoder|None, quality kind,
#:        qmin, qmax, qdefault, allowed audio)
VIDEO_FORMATS = {
    "h264": {
        "ext": "mp4", "content_type": "video/mp4",
        "encoder": "libx264", "hw_encoder": "h264_nvenc",
        "quality": "crf", "q_min": 14, "q_max": 32, "q_default": 18,
        "audio": ["aac", "none"],
    },
    "h265": {
        "ext": "mp4", "content_type": "video/mp4",
        "encoder": "libx265", "hw_encoder": "hevc_nvenc",
        "quality": "crf", "q_min": 18, "q_max": 36, "q_default": 22,
        "audio": ["aac", "none"],
    },
    "prores": {
        "ext": "mov", "content_type": "video/quicktime",
        "encoder": "prores_ks", "hw_encoder": None,
        "quality": "profile", "q_min": 0, "q_max": 5, "q_default": 3,
        "audio": ["pcm_s24le", "aac", "none"],
    },
    "vp9": {
        "ext": "webm", "content_type": "video/webm",
        "encoder": "libvpx-vp9", "hw_encoder": None,
        "quality": "crf", "q_min": 24, "q_max": 45, "q_default": 31,
        "audio": ["opus", "none"],
    },
}

AUDIO_CODECS = {
    "aac": {"bitrates": [128, 192, 256, 320], "bitrate_default": 192},
    "opus": {"bitrates": [96, 128, 192, 256], "bitrate_default": 128},
    "pcm_s24le": {"bitrates": None, "bitrate_default": None},
    "none": {"bitrates": None, "bitrate_default": None},
}

SAMPLE_RATES = [44100, 48000]

#: Delivery resolutions as the SHORT EDGE in pixels. The short edge and not the
#: height, because that is the one reading that works both ways up: 1080p is
#: ?x1080 on a landscape cut and 1080x? on a vertical one. The timeline's
#: aspect is preserved either way — this resizes the frame, it never re-frames
#: it, so a 1280x704 cut at 1080p delivers 1964x1080 and not a letterboxed
#: 1920x1080.
RESOLUTIONS = [540, 720, 1080, 1440, 2160]

DEFAULT = {
    "format": "h264",
    "quality": 18,
    "hardware": False,
    "audio": "aac",
    "audio_bitrate": 192,
    "sample_rate": 48000,
    "resolution": None,
    "fps": None,
}

#: What the intermediates are encoded at, and what every render produced before
#: this module existed. The final step can `-c:v copy` when the requested
#: output matches it exactly — see `can_copy_video`.
INTERMEDIATE = {"encoder": "libx264", "crf": 18, "pix_fmt": "yuv420p"}


def _clamp(n, lo, hi):
    return max(lo, min(hi, n))


def normalize(v):
    """Row value -> a spec that is safe to build arguments from.

    Forgiving on RANGE (clamp) and corrective on LEGALITY: an audio codec the
    container cannot mux falls back to that container's first. Passing an
    illegal pair through would fail in ffmpeg at the very end of the render,
    after all the GPU time is spent, to say that webm will not carry AAC.
    """
    v = v if isinstance(v, dict) else {}
    fmt = v.get("format") if v.get("format") in VIDEO_FORMATS else DEFAULT["format"]
    f = VIDEO_FORMATS[fmt]

    q = v.get("quality")
    q = int(round(q)) if isinstance(q, (int, float)) and not isinstance(q, bool) else f["q_default"]
    q = _clamp(q, f["q_min"], f["q_max"])

    audio = v.get("audio") if v.get("audio") in f["audio"] else f["audio"][0]
    ac = AUDIO_CODECS[audio]
    if ac["bitrates"]:
        br = v.get("audio_bitrate")
        br = int(br) if isinstance(br, (int, float)) and int(br) in ac["bitrates"] \
            else ac["bitrate_default"]
    else:
        br = 0

    sr = v.get("sample_rate")
    sr = int(sr) if sr in SAMPLE_RATES else DEFAULT["sample_rate"]

    res = v.get("resolution")
    res = int(res) if res in RESOLUTIONS else None

    fps = v.get("fps")
    fps = _clamp(int(round(fps)), 1, 120) \
        if isinstance(fps, (int, float)) and not isinstance(fps, bool) and fps > 0 else None

    return {
        "format": fmt, "quality": q,
        "hardware": bool(v.get("hardware")) and f["hw_encoder"] is not None,
        "audio": audio, "audio_bitrate": br, "sample_rate": sr,
        "resolution": res, "fps": fps,
    }


def encoder(spec):
    f = VIDEO_FORMATS[spec["format"]]
    return f["hw_encoder"] if spec["hardware"] and f["hw_encoder"] else f["encoder"]


def ext(spec):
    return VIDEO_FORMATS[spec["format"]]["ext"]


def content_type(spec):
    return VIDEO_FORMATS[spec["format"]]["content_type"]


def out_size(spec, width, height):
    """Even dimensions — yuv420p cannot represent an odd one, and ffmpeg FAILS
    the encode rather than rounding for you, on the last step of a long
    render."""
    def even(n):
        return max(2, int(round(n / 2)) * 2)
    if not spec["resolution"]:
        return even(width), even(height)
    f = spec["resolution"] / min(width, height)
    return even(width * f), even(height * f)


def can_copy_video(spec, width, height):
    """True when the requested output IS what the intermediates already are, so
    the final mux can `-c:v copy` and skip a whole re-encode of the cut.

    This is the common case — H.264 at the default CRF is the default output —
    and it is worth the check: on a 90s master the copy is seconds and the
    re-encode is minutes. Any deviation (codec, quality, hardware, resize,
    frame rate) means the pixels have to be built again.
    """
    return (spec["format"] == "h264"
            and not spec["hardware"]
            and spec["quality"] == INTERMEDIATE["crf"]
            # The SIZE, not the setting: a 1080p cut asked for 1080p needs no
            # resize either, and paying for a full re-encode to arrive at the
            # frame you already have is the kind of waste nobody would notice.
            and out_size(spec, width, height) == (width, height)
            and spec["fps"] is None)


def video_args(spec, width=None, height=None):
    """ffmpeg output args for the picture.

    NVENC does NOT take -crf: it is a different rate-control entirely, and
    passing crf to it is silently ignored rather than refused — the render
    comes back at the encoder's own default bitrate and nothing says the
    quality setting did nothing. `-cq` with `-rc vbr` is the equivalent knob.
    """
    f = VIDEO_FORMATS[spec["format"]]
    enc = encoder(spec)
    args = ["-c:v", enc]

    if spec["hardware"]:
        # -preset p5 is NVENC's balanced tier (p1 fastest .. p7 slowest); the
        # x264 preset names it also accepts are deprecated aliases.
        args += ["-rc", "vbr", "-cq", str(spec["quality"]), "-preset", "p5",
                 "-pix_fmt", "yuv420p"]
    elif f["quality"] == "profile":
        # prores_ks wants the profile as an INT and its own pixel format:
        # 422 variants are yuv422p10le, 4444 is yuva444p10le. Handing it
        # yuv420p produces a file that opens and is not ProRes-conformant.
        prof = spec["quality"]
        args += ["-profile:v", str(prof),
                 "-pix_fmt", "yuva444p10le" if prof >= 4 else "yuv422p10le"]
    elif spec["format"] == "vp9":
        # Constant-quality VP9 is -crf WITH -b:v 0; without the zero bitrate
        # libvpx treats crf as a ceiling on a bitrate-targeted encode, which is
        # a different (and much worse) mode. -row-mt is free parallelism.
        args += ["-crf", str(spec["quality"]), "-b:v", "0", "-row-mt", "1",
                 "-pix_fmt", "yuv420p"]
    else:
        args += ["-crf", str(spec["quality"]), "-preset", "veryfast",
                 "-pix_fmt", "yuv420p"]

    if spec["format"] in ("h264", "h265", "vp9") and not spec["hardware"]:
        # The intermediates are tagged bt709/tv (handlers/render.COLOR_TAGS)
        # and the copy path carries the tags for free; a re-encoding delivery
        # has to restate them or the deliverable arrives untagged and every
        # player guesses.
        args += ["-colorspace", "bt709", "-color_primaries", "bt709",
                 "-color_trc", "bt709", "-color_range", "tv"]

    if spec["fps"]:
        args += ["-r", str(spec["fps"])]
    if spec["resolution"] and width and height:
        w, h = out_size(spec, width, height)
        if (w, h) != (width, height):
            args += ["-vf", f"scale={w}:{h}:flags=lanczos"]
    # faststart is an mp4/mov concept; webm has no moov atom to move.
    if f["ext"] in ("mp4", "mov"):
        args += ["-movflags", "+faststart"]
    return args


def audio_args(spec):
    """ffmpeg output args for the sound. `none` strips the stream."""
    a = spec["audio"]
    if a == "none":
        return ["-an"]
    if a == "pcm_s24le":
        return ["-c:a", "pcm_s24le", "-ar", str(spec["sample_rate"])]
    if a == "opus":
        return ["-c:a", "libopus", "-b:a", f"{spec['audio_bitrate']}k",
                "-ar", "48000"]  # libopus only does 48k; anything else resamples anyway
    return ["-c:a", "aac", "-b:a", f"{spec['audio_bitrate']}k",
            "-ar", str(spec["sample_rate"]), "-ac", "2"]


def describe(spec):
    f = VIDEO_FORMATS[spec["format"]]
    q = (f"profile {spec['quality']}" if f["quality"] == "profile"
         else f"crf {spec['quality']}")
    bits = [spec["format"], q, encoder(spec), spec["audio"]]
    if spec["resolution"]:
        bits.append(f"{spec['resolution']}p")
    if spec["fps"]:
        bits.append(f"{spec['fps']}fps")
    return " · ".join(str(b) for b in bits)
